### PR TITLE
For #9281: Allow scanning of inverted QR codes.

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

File: components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt [hunk 1]
```diff
@@ -607,34 +607,45 @@ class QrFragment : Fragment() {
...
-            val rawResult = multiFormatReader.decodeWithState(bitmap)
-            if (rawResult != null) {
-                qrState = STATE_QRCODE_EXIST
-                rawResult.toString()
-            } else {
+        val bitmap = createBinaryBitmap(source)
+        val rawResult = try {
+            multiFormatReader.decodeWithState(bitmap)
+        } catch (e: NotFoundException) {
+            val invertedBitmap = createBinaryBitmap(source.invert())
```
Comment (csadilek) [inline components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt:628]: Right now this only tries the inverted image if a `NotFoundException` was thrown...but looking through various scanner implementations it seems possible that `decodeWithState` just returns a `null` result or throws something else entirely. I also managed to get it to crash (with an `ArrayIndexOutOfBoundsException`) by scanning repeatedly.

What do you think about this small to change to the PR which will handle all these cases? Essentially we just handle the inversion on method further up the stack.

```kotlin
 @VisibleForTesting
    internal fun tryScanningSource(source: LuminanceSource) {
        if (qrState != STATE_DECODE_PROGRESS) {
            return
        }

        scanCompleteListener?.let {
            val result = decodeSource(source) ?: decodeSource(source.invert())
            result?.let {
                scanCompleteListener?.onScanComplete(it)
            }
        }
    }

    @VisibleForTesting
    internal fun decodeSource(source: LuminanceSource): String? {
        return try {
            val bitmap = createBinaryBitmap(source)
            val rawResult = multiFormatReader.decodeWithState(bitmap)
            if (rawResult != null) {
                qrState = STATE_QRCODE_EXIST
                rawResult.toString()
            } else {
                qrState = STATE_FIND_QRCODE
                null
            }
        } catch (e: Exception) {
            qrState = STATE_FIND_QRCODE
            null
        } finally {
            multiFormatReader.reset()
        }
    }
```

**Review by csadilek — COMMENTED** [review]
Nice! This works well. Suggested a small change to handle more cases below.

File: components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt (hunk 1 above)
Comment (mcarare) [reply components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt:628]: Thanks for also testing this. I actually started with different try blocks for each decoding by the `multiFormatReader`, the second one with the inverted source only for the result being null. But with all devices I tested, I did not receive any other exception, so I simplified the code to try and decode the inverted source when the exception is thrown.

Comment (mcarare) [reply components/feature/qr/src/main/java/mozilla/components/feature/qr/QrFragment.kt]: I have made the changes and also adjusted the tests.

**Review by csadilek — APPROVED** [review]
👍 Thanks!

**Pull request merged**